substantial/types.py

```
@@ -118,17 +148,51 @@ async def exec(self, ctx, save_id, counter) -> Any:
             ctx.source(events.Event(save=save))
             return ret
         except Exception as e:
+            if strategy.compensate_on_first_fail:
+                await self._trigger_compensation(ctx, save_id, e)
+                # should be same as save fn without compensate fn ?
+                message = f"{type(e).__name__}: {e}"
+                raise RetryFail(message)
```

Ruff (0.8.2)

Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling

(B904)